lib/filewriter.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * FileWriter -- fast non-streaming file appender for accepting buffered data
+ *
+ * Locks the file when appending to be safe with multiple writers.
+ * Also, it reopens the file periodically, to make it work with logrotate.
+ *
+ * Copyright (C) 2014-2015 Andras Radics
+ * Licensed under the Apache License, Version 2.0
+ */
+
+var fs = require('fs');
+var fse = require('./fs-ext.js');
+
+var flock = fse.flock;
+var flockSync = function( fd, mode ) {
+    try { fse.flockSync(fd, mode); return true }
+    catch (err) { return false }
+}
+
+fse.SEEK_SET = 0;
+fse.SEEK_CUR = 1;
+fse.SEEK_END = 2;
+
+module.exports = FileWriter;
+
+function FileWriter( filename, openmode ) {
+    'use strict';
+
+    if (!(this instanceof FileWriter)) return new FileWriter(filename, openmode);
+    if (!filename) throw new Error("missing filename");
+
+    this._isFileWriter = true;
+    this.filename = filename;
+    this.openmode = openmode || "a";
+    // do not overwrite contents when reopening file, use a instead of w
+    this.reopenmode = this.openmode[0] !== 'w' ? openmode : ((openmode[1] === '+' || openmode[1] === 'r') ? 'a+' : 'a');
+    this.reopenInterval = 50;   // do not reuse for more than .05 seconds
+    this.reopenTime = 0;
+    this.isFirstOpen = true;
+    this.fd = undefined;
+
+    this._reopenFd = function(callback) {
+        if (this.fd !== undefined) {
+            fs.closeSync(this.fd);
+            this.fd = undefined;
+        }
+
+        var self = this;
+        var mode = this.isFirstOpen ? this.openmode : this.reopenmode;
+        // setTimeout would be faster than Date.now, but must guarantee .05 sec
+        this.reopenTime = Date.now() + this.reopenInterval;
```

Comment:
habit, I guess; I use `this` at the top level in a method, and define `self` only when
needed for a callback.  If a method refers to the object both at top-level and in callbacks,
I often use both.